### PR TITLE
feat: implement file picker capability

### PR DESCRIPTION
**Pull request by khairul-alam-licon**
# Description
- Allow Mini Apps to access files from the device using a [file input](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file). The JS SDK Sample App already has this functionality on the "File Upload" screen, so you can use the sample app to verify.
Android has `onShowFileChooser` which can be used for this purpose.

## Links
- MINI-2170

# Checklist
- [x] I have read the [contributing guidelines](../.github/CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](../.github/CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [x] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [x] I ran `./gradlew check` without errors

### CODE REVIEW

File: miniapp/USERGUIDE.md
Comment (khairul-alam-licon) [inline miniapp/USERGUIDE.md:657]: When HostApp will call `filePathHostCallback?.onReceiveValue(null)` in order to do kind of "resetting" the callback, it will be crashed if there is no exception handling in HostApp. I am not sure about this standard / idea how to mitigate this risk.

File: miniapp/USERGUIDE.md
Comment (minh-rakuten) [inline miniapp/USERGUIDE.md:643]: from what I understand is Hostapp can ignore this callback when they want to implement their own file request so you can document about it too.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt
Comment (minh-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt:180]: Is there any chance that we could get this exception?

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt
Comment (minh-rakuten) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt:257]: I think you're using `filePathHostCallback?` so you can remove `if (filePathHostCallback == null) return`
-> I'm not sure so let's test it 🙇‍♂️

**Review by minh-rakuten — COMMENTED** [review]
Looks really good to me.

@khairul-alam-licon @corycaywood Can we do anything with default implementation for file picker?

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt:180]: One chance is `ActivityNotFoundException`.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt
Comment (khairul-alam-licon) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt:257]: I think, you are right..

File: miniapp/USERGUIDE.md
Comment (khairul-alam-licon) [reply miniapp/USERGUIDE.md]: Yeah, I have wrote about it in the user guide.

File: miniapp/USERGUIDE.md
Comment (corycaywood) [inline miniapp/USERGUIDE.md:645]: I think the implementation could be simplified for the Host App: it doesn't seem like it's necessary to save the `filePathCallback` within the Host App. This could be contained within `MiniAppFilePicker`. For example, something like this:

```
val filePickingReqCode = REQUEST_CODE 
val miniAppFilePicker = MiniAppFilePicker(requestCode = filePickingReqCode)

override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    super.onActivityResult(requestCode, resultCode, data)
    if (requestCode == filePickingReqCode && resultCode == Activity.RESULT_OK) {
        data?.let { intent ->
            val result: Uri? = intent.data
            miniAppFilePicker.onReceivedFile(arrayOf(result) ?: emptyArray())
        }
    }
}
```

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFilePicker.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFilePicker.kt:16]: Related to my comment above about simplifying the integration: I don't think we need this method.

However, it would be nice if the Host App had the ability to override the `onShowFileChooser` functionality completely, so maybe we could add that here? Basically refactor your implementation from `MiniAppWebChromeClient` into this class (and then the Host App can choose to override that implementation if they want to).

```
open class MiniAppFilePicker(var requestCode: Int) {

    internal var callback: ValueCallback<Array<Uri>>? = null

    fun onShowFileChooser(
        filePathCallback: ValueCallback<Array<Uri>>?,
        fileChooserParams: WebChromeClient.FileChooserParams?,
        context: Context
    ) : Boolean {
        try {
            this.callback = filePathCallback
            val intent = fileChooserParams?.createIntent()
            (context as Activity).startActivityForResult(intent, requestCode)
        } catch (e: Exception) {
            throw (MiniAppFilePickingException(e.message.toString()))
        }
        return true
    }

    fun onReceivedFiles(files: Array<Uri>) {
        callback?.onReceiveValue(files)
    }
}
```

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt:180]: What happens with the Mini App/Host App when this exception is thrown? There's no place where the exception is caught, so isn't this going to crash the App?

Maybe we should just return an empty array to the mini app if there was an exception?

**Review by corycaywood — COMMENTED** [review]
It's a good start! I left a couple comments about how we could simplify this integration for the Host App.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFilePicker.kt
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFilePicker.kt]: Yeah, your idea is working well, specially `onReceivedFiles` is convenient for the HostApp, so I have updated it.

File: miniapp/USERGUIDE.md
Comment (khairul-alam-licon) [reply miniapp/USERGUIDE.md]: Thanks, I have updated this part.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt]: `onShowFileChooser` is expecting a `Boolean`, so I am not sure if we can return an empty array, instead maybe we can do `Log.e` the error message? 🤔

Comment (corycaywood) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt]: @khairul-alam-licon Yes, I think logging the error and then return `false` is the right solution. It's better than crashing the App 😆

Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt]: Haha, true 😄
I have updated this part accordingly...

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt:173]: Shouldn't this return the boolean from `MiniAppFilePicker`? 
```suggestion
        return miniAppFilePicker?.onShowFileChooser(filePathCallback, fileChooserParams, context) ?: false
```

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFilePicker.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFilePicker.kt:27]: I just noticed: it seems a little strange that this class is named `MiniAppFilePicker` but this method is named `onShowFileChooser`. So maybe we should either change the class name to `MiniAppFileChooser` OR change this method name to `onShowFilePicker`. What do you think?

Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFilePicker.kt]: Right, so I have renamed it to `MiniAppFileChooser`.

File: miniapp/USERGUIDE.md
Comment (corycaywood) [inline miniapp/USERGUIDE.md:660]: Is this something the Host App would really want to do? 

I think a better example here would be how they can override the `onShowFileChooser` functionality.

**Review by corycaywood — APPROVED** [review]
LGTM 👍  I'm just not sure about the `onShowFileChooser` example in the User Guide. I think we can either remove it completely or replace it with an example of how to override `onShowFileChooser`.

File: miniapp/USERGUIDE.md
Comment (khairul-alam-licon) [reply miniapp/USERGUIDE.md]: Yeah, I am thinking for an example how they can override it, currently, I have removed the guideline about it.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFileChooser.kt
Comment (minh-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFileChooser.kt:34]: Unit test for `fileChooserParams` null case.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFileChooser.kt
Comment (minh-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFileChooser.kt:49]: Unit test for `callback` null case.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt
Comment (minh-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/MiniAppWebChromeClient.kt:173]: We need to have test cases here too.

**Review by minh-rakuten — COMMENTED** [review]
Only test case job left 👍

**Review by minh-rakuten — APPROVED** [review]
It's possible to merge but let's wait until the access token get merge first.

File: miniapp/USERGUIDE.md
Comment (corycaywood) [inline miniapp/USERGUIDE.md:630]: I think these examples are in the wrong order: Most of the time the Host App will want to use `MiniAppFileChooserDefault`. So we should put that example first and make sure it's clear that it's the recommended usage.

**Pull request merged**
